src/sundials/owl_ode_sundials.ml

```diff
@@ -76,26 +70,25 @@ let cvode_s'
     ~stiff
     ~relative_tol
     ~abs_tol
-    (f : Mat.mat -> float -> Mat.mat)
+    (f : Arr.arr -> float -> Arr.arr)
     ~dt
     y0
     t0
     tstop
   =
-  let session, yvec, y, (dim1, dim2) =
-    make_cvode_session ~stiff ~relative_tol ~abs_tol f y0 t0
-  in
+  let s = Arr.shape y0 in
+  let session, yvec, y = make_cvode_session ~stiff ~relative_tol ~abs_tol f y0 t0 in
   Cvode.set_stop_time session tstop;
   fun _y t ->
     let tout = t +. dt in
     let rec until tout =
       let t', r = Cvode.solve_normal session tout yvec in
       match r with
       | Success ->
-        let y' = Mat.copy (unwrap (dim1, dim2) y) in
+        let y' = Arr.copy (Arr.reshape (unwrap y) s) in
```

Comment:
Can you explain me more about the rationale? Why is mixing up the types in this way, sometimes mat and sometimes arr, better?

Reply:
the rationale here is to support integration of higher dimensional Ndarrays (n >= 2).  `Mat.mat` and `Arr.arr` are both `genarray` underneath.